### PR TITLE
new dataset type that filters optimal replays 

### PR DESCRIPTION
**Pull request by tomsilver**

### CODE REVIEW

File: src/datasets/demo_replay.py [hunk 1]
```diff
@@ -70,8 +84,63 @@ def create_demo_replay_data(env: BaseEnv, train_tasks: List[Task]) -> Dataset:
...
+
+    The state_idx is the index into demo_traj.states where the replay
+    started.
+    """
+    assert demo_traj.states[state_idx].allclose(replay_traj.states[0])
+    # Plan starting at the end of the replay trajectory to the demo goal.
+    task = Task(replay_traj.states[-1], demo_traj.goal)
+    try:
+        policy = oracle_approach.solve(
+            task, timeout=CFG.offline_data_planning_timeout)
```
Comment (ronuchit) [inline src/datasets/demo_replay.py:116]: won't we have to change this to like 3 in order for this whole thing to actually run? maybe let's just do that now?

File: src/datasets/demo_replay.py [hunk 2]
```diff
@@ -70,8 +84,63 @@ def create_demo_replay_data(env: BaseEnv, train_tasks: List[Task]) -> Dataset:
...
+    # length of the low-level action trajectory.
+    demo_cost_to_go = _actions_to_cost_to_go(demo_traj.actions)
+    replay_actions = demo_traj.actions[:state_idx] + continued_traj.actions
+    # The +1 is for the replay itself, which consists of one option.
+    replay_cost_to_go = 1 + _actions_to_cost_to_go(replay_actions)
+    assert demo_cost_to_go <= replay_cost_to_go, "Demo was not optimal."
+    return demo_cost_to_go == replay_cost_to_go
+
+
+def _actions_to_cost_to_go(actions: Sequence[Action]) -> float:
```
Comment (ronuchit) [inline src/datasets/demo_replay.py:136]: we should probably move this to utils and call it `num_options_in_action_sequence` or something?

File: src/datasets/demo_replay.py (hunk 1 above)
Comment (ronuchit) [reply src/datasets/demo_replay.py:116]: ah okay you already did

File: tests/datasets/test_datasets.py [hunk 3]
```diff
@@ -126,3 +126,27 @@ def test_demo_replay_dataset():
...
     assert len(dataset[-1].states) == 2
     assert len(dataset[-1].actions) == 1
+
+
+def test_demo_nonoptimal_replay_dataset():
+    """Test demo+nonoptimal-replay dataset creation with Covers env."""
+    # Note that the planning timeout is intentionally set low enough that we
+    # cover some fails to plan from the replays, but not so low that planning
+    # always fails. Also the number of replays is set high enough that we
+    # consistently cover the failure case.
```
Comment (ronuchit) [inline tests/datasets/test_datasets.py:136]: which failure case, do you mean ApproachFailure

File: tests/datasets/test_datasets.py [hunk 4]
```diff
@@ -126,3 +126,27 @@ def test_demo_replay_dataset():
     dataset = create_dataset(env, train_tasks)
     assert len(dataset[-1].states) == 2
     assert len(dataset[-1].actions) == 1
+
+
+def test_demo_nonoptimal_replay_dataset():
+    """Test demo+nonoptimal-replay dataset creation with Covers env."""
+    # Note that the planning timeout is intentionally set low enough that we
+    # cover some fails to plan from the replays, but not so low that planning
```
Comment (ronuchit) [inline tests/datasets/test_datasets.py:134]: failures*

File: tests/datasets/test_datasets.py (hunk 3 above)
Comment (ronuchit) [reply tests/datasets/test_datasets.py:136]: @tomsilver question for you

Comment (tomsilver) [reply tests/datasets/test_datasets.py:136]: yeah ApproachFailure or ApproachTimeout

File: src/datasets/demo_replay.py (hunk 2 above)
Comment (tomsilver) [reply src/datasets/demo_replay.py:136]: sounds good

**Review by ronuchit — APPROVED** [review]

**Pull request merged**